tests/transforms/test_batch_transform.py

```diff
@@ -0,0 +1,427 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for the batch transform.
+"""
+
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+class TestBatchTransform:
+    """Unit tests for the batch_transform class"""
+
+    def test_error_invalid_callable(self):
+        """Test that an error is raised if the transform
+        is applied to an invalid function"""
+
+        with pytest.raises(ValueError, match="does not appear to be a valid Python function"):
+            qml.batch_transform(5)
+
+    def test_none_processing(self):
+        """Test that a transform that return None for a processing function applies
+        the identity as the processing function"""
+
+        @qml.batch_transform
+        def my_transform(tape):
+            tape1 = tape.copy()
+            tape2 = tape.copy()
+            return [tape1, tape2], None
+
+        with qml.tape.QuantumTape() as tape:
+            qml.Hadamard(wires=0)
+            qml.expval(qml.PauliX(0))
+
+        tapes, fn = my_transform(tape)
+        assert fn(5) == 5
+
+    def test_not_differentiable(self):
+        """Test that a non-differentiable transform cannot be differentiated"""
+
+        def my_transform(tape):
+            tape1 = tape.copy()
+            tape2 = tape.copy()
+            return [tape1, tape2], qml.math.sum
+
+        my_transform = qml.batch_transform(my_transform, differentiable=False)
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @my_transform
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.Hadamard(wires=0)
+            qml.RY(x, wires=0)
+            return qml.expval(qml.PauliX(0))
+
+        res = circuit(0.5)
+        assert isinstance(res, float)
+        assert not np.allclose(res, 0)
+
+        with pytest.warns(UserWarning, match="Output seems independent of input"):
+            qml.grad(circuit)(0.5)
+
+    def test_expand_fn(self, mocker):
+        """Test that if an expansion function is provided,
+        that the input tape is expanded before being transformed."""
+
+        def expand_fn(tape):
+            return tape.expand(stop_at=lambda obj: obj.name != "PhaseShift")
+
+        class MyTransform:
+            """Dummy class to allow spying to work"""

```

Comment:
Nice test :grin:

Reply:
This single test took like 90% of the test writing time haha. Took so long to work out how to get spying to work.